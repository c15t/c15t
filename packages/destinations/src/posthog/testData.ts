// cSpell:disable
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
//@ts-nocheck
import type {
	ActionDefinition,
	DestinationDefinition,
	InputField,
} from '@segment/actions-core';

import Chance from 'chance';

type Choices = Pick<InputField, 'choices'>;

function setTestData(
	seedName: string,
	type: string,
	fieldName?: string,
	format?: string,
	choices?: Choices
) {
	const chance = new Chance(seedName);

	if (Array.isArray(choices)) {
		const choice = chance.pickone(choices);

		if (choice.value) {
			return choice.value;
		}
		return choice;
	}
	let val: any;
	switch (type) {
		case 'boolean':
			val = chance.bool();
			break;
		case 'datetime':
			val = '2021-02-01T00:00:00.000Z';
			break;
		case 'integer':
			val = chance.integer();
			break;
		case 'number':
			val = chance.floating({ fixed: 2 });
			break;
		case 'object':
			val = { testType: chance.string() };
			break;
		case 'text':
			val = chance.sentence();
			break;
		default:
			// covers string
			switch (format) {
				case 'date': {
					val = '2021-02-01T00:00:00.000Z';
					break;
				}
				case 'date-time':
					val = '2021-02-01T00:00:00.000Z';
					break;
				case 'email':
					val = chance.email();
					break;
				case 'hostname':
					val = chance.domain();
					break;
				case 'ipv4':
					val = chance.ip();
					break;
				case 'ipv6':
					val = chance.ipv6();
					break;
				case 'time': {
					const d = chance.date();
					val = [d.getHours(), d.getMinutes(), d.getSeconds()]
						.map((v) => String(v).padStart(2, '0'))
						.join(':');
					break;
				}
				case 'uri':
					val = chance.url();
					break;
				case 'uuid':
					val = chance.guid();
					break;
				default:
					val = chance.string();
					break;
			}
			break;
	}

	if (fieldName === 'email') {
		val = chance.email();
	}
	if (fieldName === 'currency') {
		val = chance.currency().code;
	}

	return val;
}

function setData(
	eventData: any,
	chanceName: string,
	fieldName: string,
	field: any,
	data?: any
) {
	const { format, multiple, type } = field;

	if (!data) {
		// biome-ignore lint/style/noParameterAssign: This is a generated file
		data = setTestData(chanceName, type, fieldName, format, field.choices);
	}

	eventData[fieldName] = multiple ? [data] : data;
	return eventData;
}

export function generateTestData(
	seedName: string,
	destination: DestinationDefinition<any>,
	action: ActionDefinition<any>,
	isRequiredOnly: boolean
) {
	let eventData: any = {};
	const settingsData: any = {};

	const authentication = destination.authentication;
	if (authentication) {
		for (const settingKey in authentication.fields) {
			const { format, type } = authentication.fields[settingKey];
			settingsData[settingKey] = setTestData(seedName, type, undefined, format);
		}
	}

	for (const [name, field] of Object.entries(action.fields)) {
		if (isRequiredOnly && !(field.required || name.includes('id'))) {
			continue;
		}

		const { properties } = field;

		if (properties) {
			let subData: any = {};
			let propertyFields = Object.keys(properties);

			if (isRequiredOnly) {
				propertyFields = propertyFields.filter(
					(name) => properties[name].required
				);
			}

			for (const propertyName of propertyFields) {
				const property = properties[propertyName];
				subData = setData(subData, seedName, propertyName, property);
			}

			eventData = setData(eventData, seedName, name, field, subData);
			continue;
		}

		eventData = setData(eventData, seedName, name, field);
	}

	return [eventData, settingsData];
}
