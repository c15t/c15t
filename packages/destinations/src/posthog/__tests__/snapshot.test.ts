import { createTestEvent, createTestIntegration } from '@segment/actions-core';
import nock from 'nock';
import { describe, expect, it } from 'vitest';

import destination from '../index';
import { generateTestData } from '../testData';

const testDestination = createTestIntegration(destination);
const destinationSlug = 'post-hog';

describe(`Testing snapshot for ${destinationSlug} destination:`, () => {
	for (const actionSlug in destination.actions) {
		it(`${actionSlug} action - required fields`, async () => {
			const seedName = `${destinationSlug}#${actionSlug}`;
			const action = destination.actions[actionSlug];

			if (!action) {
				return null;
			}

			const [eventData, settingsData] = generateTestData(
				seedName,
				destination,
				action,
				true
			);

			nock(/.*/).persist().get(/.*/).reply(200);
			nock(/.*/).persist().post(/.*/).reply(200);
			nock(/.*/).persist().put(/.*/).reply(200);

			const event = createTestEvent({
				properties: eventData,
			});

			const responses = await testDestination.testAction(actionSlug, {
				auth: undefined,
				event: event,
				mapping: event.properties,
				settings: settingsData,
			});

			if (!responses[0]) {
				return null;
			}

			const request = responses[0].request;
			const rawBody = await request.text();

			try {
				const json = JSON.parse(rawBody);
				expect(json).toMatchSnapshot({
					timestamp: expect.any(String),
				});
				return;
			} catch (_err) {
				expect(rawBody).toMatchSnapshot();
			}

			expect(request.headers).toMatchSnapshot();
		});

		it(`${actionSlug} action - all fields`, async () => {
			const seedName = `${destinationSlug}#${actionSlug}`;
			const action = destination.actions[actionSlug];
			if (!action) {
				return null;
			}
			const [eventData, settingsData] = generateTestData(
				seedName,
				destination,
				action,
				false
			);

			nock(/.*/).persist().get(/.*/).reply(200);
			nock(/.*/).persist().post(/.*/).reply(200);
			nock(/.*/).persist().put(/.*/).reply(200);

			const event = createTestEvent({
				properties: eventData,
			});

			const responses = await testDestination.testAction(actionSlug, {
				auth: undefined,
				event: event,
				mapping: event.properties,
				settings: settingsData,
			});

			if (!responses[0]) {
				return null;
			}

			const request = responses[0].request;
			const rawBody = await request.text();

			try {
				const json = JSON.parse(rawBody);
				expect(json).toMatchSnapshot({
					timestamp: expect.any(String),
				});
				return;
			} catch (_err) {
				expect(rawBody).toMatchSnapshot();
			}
		});
	}
});
