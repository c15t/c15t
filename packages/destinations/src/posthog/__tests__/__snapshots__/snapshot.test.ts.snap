// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing snapshot for posthog destination: > group action - all fields 1`] = `
{
  "api_key": "nJ29t#",
  "distinct_id": "nJ29t#",
  "event": "$groupidentify",
  "headers": {},
  "properties": {
    "$group_key": "nJ29t#",
    "$group_set": {
      "date_joined": "2021-02-01T00:00:00.000Z",
      "name": "nJ29t#",
      "testType": "nJ29t#",
    },
    "$group_type": "organization",
    "testType": "nJ29t#",
  },
  "timestamp": Any<String>,
}
`;

exports[`Testing snapshot for posthog destination: > group action - required fields 1`] = `
{
  "api_key": "nJ29t#",
  "distinct_id": "nJ29t#",
  "event": "$groupidentify",
  "headers": {},
  "properties": {
    "$group_key": "nJ29t#",
    "$group_set": {
      "date_joined": "2021-02-01T00:00:00.000Z",
      "name": "nJ29t#",
    },
    "$group_type": "organization",
  },
  "timestamp": Any<String>,
}
`;

exports[`Testing snapshot for posthog destination: > identify action - all fields 1`] = `
{
  "api_key": "!jScCo^zqCWl8wN",
  "distinct_id": "!jScCo^zqCWl8wN",
  "event": "$identify",
  "headers": {},
  "properties": {
    "$set": {
      "testType": "!jScCo^zqCWl8wN",
    },
  },
  "timestamp": Any<String>,
}
`;

exports[`Testing snapshot for posthog destination: > identify action - required fields 1`] = `
{
  "api_key": "!jScCo^zqCWl8wN",
  "distinct_id": "!jScCo^zqCWl8wN",
  "event": "$identify",
  "headers": {},
  "properties": {},
  "timestamp": Any<String>,
}
`;

exports[`Testing snapshot for posthog destination: > page action - all fields 1`] = `
{
  "api_key": "$]bA$h4w!R%Q]!1O",
  "distinct_id": "$]bA$h4w!R%Q]!1O",
  "event": "$pageview",
  "headers": {},
  "properties": {
    "$browser": "$]bA$h4w!R%Q]!1O",
    "$browser_language": "$]bA$h4w!R%Q]!1O",
    "$browser_version": 39298047640862.72,
    "$current_url": "$]bA$h4w!R%Q]!1O",
    "$device_type": "$]bA$h4w!R%Q]!1O",
    "$host": "$]bA$h4w!R%Q]!1O",
    "$insert_id": "$]bA$h4w!R%Q]!1O",
    "$ip": "$]bA$h4w!R%Q]!1O",
    "$lib": "$]bA$h4w!R%Q]!1O",
    "$lib_version": "$]bA$h4w!R%Q]!1O",
    "$os": "$]bA$h4w!R%Q]!1O",
    "$os_version": "$]bA$h4w!R%Q]!1O",
    "$pathname": "$]bA$h4w!R%Q]!1O",
    "$raw_user_agent": "$]bA$h4w!R%Q]!1O",
    "$referrer": "$]bA$h4w!R%Q]!1O",
    "$referring_domain": "$]bA$h4w!R%Q]!1O",
    "$screen_dpr": 39298047640862.72,
    "$screen_height": 39298047640862.72,
    "$screen_width": 39298047640862.72,
    "$session_id": "$]bA$h4w!R%Q]!1O",
    "$time": 39298047640862.72,
    "$viewport_height": 39298047640862.72,
    "$viewport_width": 39298047640862.72,
    "$window_id": "$]bA$h4w!R%Q]!1O",
    "title": "$]bA$h4w!R%Q]!1O",
    "utm_campaign": "$]bA$h4w!R%Q]!1O",
    "utm_content": "$]bA$h4w!R%Q]!1O",
    "utm_medium": "$]bA$h4w!R%Q]!1O",
    "utm_source": "$]bA$h4w!R%Q]!1O",
    "utm_term": "$]bA$h4w!R%Q]!1O",
  },
  "timestamp": Any<String>,
}
`;

exports[`Testing snapshot for posthog destination: > page action - required fields 1`] = `
{
  "api_key": "$]bA$h4w!R%Q]!1O",
  "distinct_id": "$]bA$h4w!R%Q]!1O",
  "event": "$pageview",
  "headers": {},
  "properties": {
    "$insert_id": "$]bA$h4w!R%Q]!1O",
    "$screen_width": 39298047640862.72,
    "$session_id": "$]bA$h4w!R%Q]!1O",
    "$viewport_width": 39298047640862.72,
    "$window_id": "$]bA$h4w!R%Q]!1O",
  },
  "timestamp": Any<String>,
}
`;

exports[`Testing snapshot for posthog destination: > track action - all fields 1`] = `
{
  "api_key": "hQwZ!",
  "distinct_id": "hQwZ!",
  "event": "hQwZ!",
  "headers": {},
  "properties": {
    "$browser": "hQwZ!",
    "$browser_language": "hQwZ!",
    "$browser_version": -86879145568501.77,
    "$current_url": "hQwZ!",
    "$device_type": "hQwZ!",
    "$host": "hQwZ!",
    "$insert_id": "hQwZ!",
    "$ip": "hQwZ!",
    "$lib": "hQwZ!",
    "$lib_version": "hQwZ!",
    "$os": "hQwZ!",
    "$os_version": "hQwZ!",
    "$pathname": "hQwZ!",
    "$raw_user_agent": "hQwZ!",
    "$referrer": "hQwZ!",
    "$referring_domain": "hQwZ!",
    "$screen_dpr": -86879145568501.77,
    "$screen_height": -86879145568501.77,
    "$screen_width": -86879145568501.77,
    "$session_id": "hQwZ!",
    "$time": -86879145568501.77,
    "$viewport_height": -86879145568501.77,
    "$viewport_width": -86879145568501.77,
    "$window_id": "hQwZ!",
    "distinct_id": "hQwZ!",
    "testType": "hQwZ!",
    "title": "hQwZ!",
    "utm_campaign": "hQwZ!",
    "utm_content": "hQwZ!",
    "utm_medium": "hQwZ!",
    "utm_source": "hQwZ!",
    "utm_term": "hQwZ!",
  },
  "timestamp": Any<String>,
}
`;

exports[`Testing snapshot for posthog destination: > track action - required fields 1`] = `
{
  "api_key": "hQwZ!",
  "distinct_id": "hQwZ!",
  "event": "hQwZ!",
  "headers": {},
  "properties": {
    "$insert_id": "hQwZ!",
    "$screen_width": -86879145568501.77,
    "$session_id": "hQwZ!",
    "$viewport_width": -86879145568501.77,
    "$window_id": "hQwZ!",
    "distinct_id": "hQwZ!",
  },
  "timestamp": Any<String>,
}
`;
